### PR TITLE
Add ability for TQDMProgressBar to retain prior epoch training bars

### LINKED ISSUE #19322: stats logging in "on_train_epoch_end" ends up on wrong progress bar
### Bug description

When logging statistics at the end of an epoch from within `on_train_epoch_end`, the statistics end up on the wrong progress bar.

Since there doesn't seem to be a configuration to tell lightning nor the `TQDMProgressBar` to retain the bar for each epoch, I've been forced to inject a new line after each epoch ends, in order to not lose any of the valuable statistics in the console output.

The following is the output from a 3 epoch run:
```
Epoch 0: 100%|█████████████████████████| 938/938 [00:04<00:00, 206.06it/s, v_num=207]
Epoch 1: 100%|█████████████| 938/938 [00:04<00:00, 233.29it/s, v_num=207, loss=0.553]
Epoch 2: 100%|█████████████| 938/938 [00:04<00:00, 233.39it/s, v_num=207, loss=0.329]
`Trainer.fit` stopped: `max_epochs=3` reached.
Epoch 2: 100%|█████████████| 938/938 [00:04<00:00, 232.93it/s, v_num=207, loss=0.329]
```
* Loss for epoch 0 is incorrectly shown for epoch 1. 
* Loss for epoch 1 is incorrectly shown for epoch 2. 
* No logged loss at all is reported for epoch 0 nor epoch 2, 

If there is a _proper_ way to retain the progress bar for each epoch that is different from what I'm doing, then please let me know and this ticket can then be closed. If not, hopefully a fix can be found.

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

```python
import torch
import torchvision
import pytorch_lightning as pl

class DemoNet(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.fc = torch.nn.Linear(784, 10)
        self.batch_losses = []

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.001)

    def training_step(self, batch:torch.Tensor, _):
        x, y = batch
        x = x.reshape(x.size(0), -1)
        yh = self.fc(x)
        loss = torch.nn.functional.cross_entropy(yh, y)
        self.batch_losses.append(loss)
        return loss

    def on_train_epoch_end(self):
        loss = torch.stack(self.batch_losses).mean()
        self.log('loss', loss, on_step=False, on_epoch=True, prog_bar=True)
        self.batch_losses.clear()
        print("")

ds = torchvision.datasets.MNIST(root="dataset/", train=True, transform=torchvision.transforms.ToTensor(), download=True)
train_loader = torch.utils.data.DataLoader(dataset=ds, batch_size=64, shuffle=False)
trainer = pl.Trainer(max_epochs=3)
trainer.fit(DemoNet(), train_loader)
```

### Error messages and logs

N/A

### Environment

<details>
  <summary>Current environment</summary>

* Lightning:
	- lightning-utilities: 0.9.0
	- pytorch-lightning: 2.1.3
	- torch:             2.1.2+cu118
	- torchaudio:        2.1.2+cu118
	- torchmetrics:      1.2.1
	- torchvision:       0.16.2+cu118
	- tqdm:              4.66.1
* System:
	- OS:                Windows
	- architecture:
		- 64bit
		- WindowsPE
	- processor:         AMD64 Family 25 Model 97 Stepping 2, AuthenticAMD
	- python:            3.10.0
	- release:           10
	- version:           10.0.19045
* CUDA:
	- GPU:
		- NVIDIA GeForce RTX 4090
	- available:         True
	- version:           11.8

*  How you installed Lightning(`conda`, `pip`, source): pip
*  Running environment of LightningApp (e.g. local, cloud): local

</details>

### More info

_No response_

cc @carmocca

### PR DESCRIPTION
**Pull request by jojje**
## What does this PR do?
Adds the ability to retain tqdm progress bars from prior training epochs.

Fixes #19322

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs).  
   A: No, an issue was was opened but there is no PR yet to discuss (catch 22).
- [ X ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ X ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
   A: Yes.
- Did you write any **new necessary tests**? (not for typos and docs)
   A: No. The goal was to _not_ break any existing behavior so the existing TQDMProgressBar test suite was used to ensure that objective.
- [ X ] Did you verify new and **existing tests pass** locally with your changes?
   A: Partly, I only ran the TQDMProgressBar test suite, since the other's mostly failed on my machine even before the change was made. As such am submitting a draft PR to ensure the entire test suite passes using the project's CI system before it's worth discussing the PR.
- Did you list all the **breaking changes** introduced by this pull request? 
   A: N/A, No breaking changes intended.
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors): 
   A: Yes

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19578.org.readthedocs.build/en/19578/

### CODE REVIEW

File: src/lightning/pytorch/callbacks/progress/tqdm_progress.py [hunk 1]
```diff
@@ -99,12 +99,13 @@ class TQDMProgressBar(ProgressBar):
...
             :paramref:`~lightning.pytorch.trainer.trainer.Trainer.process_position` in the
             :class:`~lightning.pytorch.trainer.trainer.Trainer`.
+        leave: Leaves the finished progress bar in the terminal at the end of the epoch. Default: False
 
     """
 
     BAR_FORMAT = "{l_bar}{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}, {rate_noinv_fmt}{postfix}]"
 
-    def __init__(self, refresh_rate: int = 1, process_position: int = 0):
+    def __init__(self, refresh_rate: int = 1, process_position: int = 0, leave: bool = False):
```
Comment (awaelchli) [inline src/lightning/pytorch/callbacks/progress/tqdm_progress.py:108]: Sounds like a good idea!

File: src/lightning/pytorch/callbacks/progress/tqdm_progress.py [hunk 2]
```diff
@@ -262,6 +264,8 @@ def on_train_start(self, *_: Any) -> None:
 
     @override
     def on_train_epoch_start(self, trainer: "pl.Trainer", *_: Any) -> None:
+        if self._leave:
+            self.train_progress_bar = self.init_train_tqdm()
```
Comment (awaelchli) [inline src/lightning/pytorch/callbacks/progress/tqdm_progress.py:268]: Instead of doing this, I believe it would be better to pass `leave=self.leave` to the tqdm bar directly (see `init_train_tqdm()`, `init_validation_tqdm()` etc. above.

Comment (awaelchli) [reply src/lightning/pytorch/callbacks/progress/tqdm_progress.py:268]: No, I'm saying simply pass `leave=self.leave` in `init_train_tqdm()` where we hard coded it to True so far. Line 205

**Review by awaelchli — APPROVED** [review]
